### PR TITLE
add new api for roaring bitmap and  add api example in README.md

### PR DESCRIPTION
**Pull request by baiyunping333**

### CODE REVIEW

**Comment by aviggiano** [comment]
Hey thank you very much for this contribution, I love it.
I'll take a closer look at it over the weekend, but I think these commands and documentation will be very useful.

File: CMakeLists.txt [hunk 1]
```diff
@@ -46,6 +46,7 @@ set(CMAKE_POSITION_INDEPENDENT_CODE ON)
 
 set(SOURCE_FILES ${SRC_PATH}/redis-roaring.c
         ${SRC_PATH}/type.c
-        ${SRC_PATH}/data-structure.c)
+        ${SRC_PATH}/data-structure.c
+        ${SRC_PATH}/sds.c)
```
Comment (aviggiano) [inline CMakeLists.txt]: Maybe sds could be an external dependency as a git module, just like `CRoaring` and `hiredis` are?
https://github.com/antirez/sds

File: README.md [hunk 2]
```diff
@@ -19,34 +20,22 @@ Pull requests are welcome.
 
 ## Getting started
 
-Clone this repository:
-
-```bash
+```
 $ git clone https://github.com/aviggiano/redis-roaring.git
 $ cd redis-roaring/
+$ configure.sh
```
Comment (aviggiano) [inline README.md]: Could you also add these steps into the automated setup script?

https://github.com/aviggiano/redis-roaring/blob/master/configure.sh

File: README.md [hunk 3]
```diff
@@ -76,10 +65,39 @@ Currently the following operations are supported
...
+127.0.0.1:6379> r.setrange test 1 100
+
+# get all the int numbers
+127.0.0.1:6379> R.GETINTARRAY test
+
+# fill up the roaring bitmap, then don't use the `R.GETINTARRAY` 
+# because you need 2^32*4bytes memory and a long lang time
+127.0.0.1:6379> R.SETFULL test
+
+# bu you can use `R.RANGEINTARRAY` to get numbers from 100 to 1000
```
Comment (aviggiano) [inline README.md]: but

File: src/data-structure.c [hunk 4]
```diff
@@ -106,6 +106,12 @@ uint32_t* bitmap_get_int_array(Bitmap* bitmap, size_t* n) {
   roaring_bitmap_to_uint32_array(bitmap, ans);
   return ans;
 }
+uint32_t* bitmap_range_int_array(Bitmap* bitmap, size_t offset,  size_t n) {
+  uint32_t* ans = malloc(sizeof(*ans) * n);
+  memset(ans, 0, sizeof(*ans) * n);
```
Comment (aviggiano) [inline src/data-structure.c]: Maybe these two lines could be a `calloc`?

```C
uint32_t* ans = calloc(n, sizeof(*ans));
```

File: src/redis-roaring.c [hunk 5]
```diff
@@ -1,11 +1,82 @@
...
+  RedisModuleKey* key = RedisModule_OpenKey(ctx, argv[1], REDISMODULE_READ | REDISMODULE_WRITE);
+  int type = RedisModule_KeyType(key);
+  if(type != REDISMODULE_KEYTYPE_EMPTY){
+    return RedisModule_ReplyWithError(ctx, "key exists");
+  }
+
+  Bitmap* nbitmap = roaring_bitmap_from_range(0, UINT32_MAX-1, 1);
+  RedisModule_ModuleTypeSetValue(key, BitmapType, nbitmap);
+  RedisModule_ReplicateVerbatim(ctx);
+  RedisModule_ReplyWithSimpleString(ctx, "ok");
```
Comment (aviggiano) [inline src/redis-roaring.c]: "OK" in caps, to mimic redis' native API

File: src/redis-roaring.c [hunk 6]
```diff
@@ -82,6 +153,76 @@ int RGetBitCommand(RedisModuleCtx* ctx, RedisModuleString** argv, int argc) {
 
   return REDISMODULE_OK;
 }
+/**
+ * R.STAT <key> 
+ * */
+int RStatBitCommand(RedisModuleCtx* ctx, RedisModuleString** argv, int argc) {
```
Comment (aviggiano) [inline src/redis-roaring.c]: Nice command

File: src/redis-roaring.c [hunk 7]
```diff
@@ -82,6 +153,76 @@ int RGetBitCommand(RedisModuleCtx* ctx, RedisModuleString** argv, int argc) {
...
+  RedisModule_ReplicateVerbatim(ctx);
+  RedisModule_ReplyWithSimpleString(ctx, s);    
+  sdsfree(s);
+
+  return REDISMODULE_OK;
+}
+/**
+ * R.OPTIMIZE <key> 
+ * */
+int ROptimizeBitCommand(RedisModuleCtx* ctx, RedisModuleString** argv, int argc) {
```
Comment (aviggiano) [inline src/redis-roaring.c]: Also very useful

File: src/data-structure.c (hunk 4 above)
Comment (baiyunping333) [reply src/data-structure.c]: very thanks

**Comment by baiyunping333** [comment]
@aviggiano  hi, how about the merge?

**Comment by aviggiano** [comment]
@baiyunping333 I'll do it today!

**Comment by aviggiano** [comment]
Sorry @baiyunping333 I accidentally closed the PR while trying to update the documentation. Could you re-open it please?

My bad.

**Pull request closed without merging**